ctera/src/main/java/ch/cyberduck/core/ctera/CteraSession.java

```diff
@@ -178,9 +272,74 @@ public <T> T _getFeature(final Class<T> type) {
         if(type == UrlProvider.class) {
             return (T) new CteraUrlProvider(host);
         }
+        if(type == AttributesFinder.class) {
+            return (T) attributes;
+        }
+        if(type == ListService.class) {
+            return (T) list;
+        }
+        if(type == AclPermission.class) {
+            return (T) new CteraAclPermissionFeature(this);
+        }
+        if(type == Read.class) {
+            return (T) new DAVReadFeature(this) {
+                @Override
+                public void preflight(Path file) throws BackgroundException {
+                    // TODO CTERA-136 do we need both? Which one to prefer?
+                    super.preflight(file);
+                    session.checkCteraRole(file, READPERMISSION);
+                }
+            };
+        }
+        if(type == Write.class) {
+            return (T) new DAVWriteFeature(this) {
+                @Override
+                public void preflight(Path file) throws BackgroundException {
+                    // TODO CTERA-136 do we need both? Which one to prefer?
+                    super.preflight(file);
+                    session.checkCteraRole(file, WRITEPERMISSION);
+                }
+            };
+        }
+        if(type == Delete.class) {
+            return (T) new DAVDeleteFeature(this) {
+                @Override
+                public void preflight(Path file) throws BackgroundException {
+                    // TODO CTERA-136 do we require writepermission on file as well?
+                    super.preflight(file);
+                    session.checkCteraRole(file, DELETEPERMISSION);
+                }
+            };
+        }
+        if(type == Copy.class) {
+            return (T) new DAVCopyFeature(this) {
+                @Override
+                public void preflight(final Path source, final Path target) throws BackgroundException {
+                    // TODO CTERA-136 do we require writepermission on target's parent?
+                    super.preflight(source, target);
+                    if(source.isDirectory()) {
+                        session.checkCteraRole(target.getParent(), CREATEDIRECTORIESPERMISSION);
+                    }
+                    else {
+                        session.checkCteraRole(target.getParent(), CREATEFILEPERMISSION);
+                    }
+                }
+            };
+        }
         return super._getFeature(type);
     }
 
+    protected void checkCteraRole(final Path file, final Acl.Role role) throws BackgroundException {
```

Comment:
Move to `CteraAclPermissionFeature`.

Reply:
Done 652832da5fcb764138567e340893638e9846b71f